[JSC] Handle `String.prototype.includes` in DFG/FTL

sosukesuzuki opened the pull request:
#### cc5da5a661a3dda8b8edd19f903fed3de7c35f4b
<pre>
[JSC] Handle `String.prototype.includes` in DFG/FTL
<a href="https://bugs.webkit.org/show_bug.cgi?id=304138">https://bugs.webkit.org/show_bug.cgi?id=304138</a>

Reviewed by Yusuke Suzuki.

This patch adds support for `String.prototype.includes` in DFG/FTL using C++
operations like `String.prototype.indexOf`.

This patch makes `String.prototype.includes` 1.2x to 1.4x faster in the following
microbenchmarks. Furthermore, it improves even more significantly when folding is
effective.

                                                  TipOfTree                  Patched

string-prototype-includes-with-index-with-one-char
                                               74.6933+-36.5188          65.5386+-26.0564         might be 1.1397x faster
string-prototype-includes                      70.8373+-31.3757          69.8322+-33.4859         might be 1.0144x faster
string-prototype-includes-with-index          113.6356+-70.8584          73.4337+-24.0293         might be 1.5475x faster
string-prototype-includes-constant-folding
                                               84.4210+-17.2479    ^      8.6475+-4.8715        ^ definitely 9.7625x faster
string-prototype-includes-with-one-char        52.5742+-20.4285          36.3516+-12.4097         might be 1.4463x faster

Tests: JSTests/stress/string-includes.js

* JSTests/microbenchmarks/string-prototype-includes-constant-folding.js: Added.
(testConstantFound):
(testConstantNotFound):
(testConstantWithIndex):
(testConstantOneChar):
(testConstantOneCharWithIndex):
* JSTests/microbenchmarks/string-prototype-includes-with-index-with-one-char.js: Added.
(test):
(makeString):
* JSTests/microbenchmarks/string-prototype-includes-with-index.js: Added.
(test):
(makeString):
* JSTests/microbenchmarks/string-prototype-includes-with-one-char.js: Added.
(test):
(makeString):
* JSTests/microbenchmarks/string-prototype-includes.js: Added.
(test):
(makeString):
* JSTests/stress/string-prototype-includes-constant-folding.js: Added.
(shouldBe):
(testConstantFound):
(testConstantNotFound):
(testConstantWithIndex):
(testConstantWithIndexNotFound):
(testConstantOneChar):
(testConstantOneCharNotFound):
(testConstantOneCharWithIndex):
(testConstantOneCharWithIndexNotFound):
* JSTests/stress/string-prototype-includes-with-index-with-one-char.js: Added.
(shouldBe):
(test):
(makeString):
* JSTests/stress/string-prototype-includes-with-index.js: Added.
(shouldBe):
(test):
(makeString):
* JSTests/stress/string-prototype-includes-with-one-char.js: Added.
(shouldBe):
(test):
(makeString):
* JSTests/stress/string-prototype-includes.js: Added.
(shouldBe):
(test):
(makeString):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):

Canonical link: <a href="https://commits.webkit.org/304852@main">https://commits.webkit.org/304852@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/daa18f4c8863f39fc893ada23adab41919bcddc6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136667 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9026 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47954 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144392 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89640 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9727 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8872 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104510 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139612 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7103 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122457 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85349 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6747 "Passed tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4437 "Found 1 new API test failure: TestWebKitAPI.WKWebExtensionDataRecord.GetDataRecordsForMultipleContexts (failure)") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4987 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128628 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116063 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40646 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147151 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/135153 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8710 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41219 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112865 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8728 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7330 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113194 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28752 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6672 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118749 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62829 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8758 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36803 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167932 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8479 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72324 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43817 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8698 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8550 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (commented):
How about converting it to StringIndexOf(...) !== -1? So we do not need to have similar nodes twice, and StringIndexOf can be CSE-ed

sosukesuzuki committed:
[JSC] Handle `String.prototype.includes` in DFG/FTL
https://bugs.webkit.org/show_bug.cgi?id=304138

Reviewed by Yusuke Suzuki.

This patch adds support for `String.prototype.includes` in DFG/FTL using C++
operations like `String.prototype.indexOf`.

This patch makes `String.prototype.includes` 1.2x to 1.4x faster in the following
microbenchmarks. Furthermore, it improves even more significantly when folding is
effective.

                                                  TipOfTree                  Patched

string-prototype-includes-with-index-with-one-char
                                               74.6933+-36.5188          65.5386+-26.0564         might be 1.1397x faster
string-prototype-includes                      70.8373+-31.3757          69.8322+-33.4859         might be 1.0144x faster
string-prototype-includes-with-index          113.6356+-70.8584          73.4337+-24.0293         might be 1.5475x faster
string-prototype-includes-constant-folding
                                               84.4210+-17.2479    ^      8.6475+-4.8715        ^ definitely 9.7625x faster
string-prototype-includes-with-one-char        52.5742+-20.4285          36.3516+-12.4097         might be 1.4463x faster

Tests: JSTests/stress/string-includes.js

* JSTests/microbenchmarks/string-prototype-includes-constant-folding.js: Added.
(testConstantFound):
(testConstantNotFound):
(testConstantWithIndex):
(testConstantOneChar):
(testConstantOneCharWithIndex):
* JSTests/microbenchmarks/string-prototype-includes-with-index-with-one-char.js: Added.
(test):
(makeString):
* JSTests/microbenchmarks/string-prototype-includes-with-index.js: Added.
(test):
(makeString):
* JSTests/microbenchmarks/string-prototype-includes-with-one-char.js: Added.
(test):
(makeString):
* JSTests/microbenchmarks/string-prototype-includes.js: Added.
(test):
(makeString):
* JSTests/stress/string-prototype-includes-constant-folding.js: Added.
(shouldBe):
(testConstantFound):
(testConstantNotFound):
(testConstantWithIndex):
(testConstantWithIndexNotFound):
(testConstantOneChar):
(testConstantOneCharNotFound):
(testConstantOneCharWithIndex):
(testConstantOneCharWithIndexNotFound):
* JSTests/stress/string-prototype-includes-with-index-with-one-char.js: Added.
(shouldBe):
(test):
(makeString):
* JSTests/stress/string-prototype-includes-with-index.js: Added.
(shouldBe):
(test):
(makeString):
* JSTests/stress/string-prototype-includes-with-one-char.js: Added.
(shouldBe):
(test):
(makeString):
* JSTests/stress/string-prototype-includes.js: Added.
(shouldBe):
(test):
(makeString):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):

Canonical link: https://commits.webkit.org/304852@main
